### PR TITLE
Use postprocessors for AccumBounds instead of hard-coding it in Mul/Add.flatten

### LINKED ISSUE #13159: Mul(AccumBounds(-1, 1), 0) gives AccumBounds(-1, 1)
```py
>>> Mul(AccumBounds(-1, 1), 0)
<-1, 1>
```

Even though

```py
>>> Mul(0, AccumBounds(-1, 1))
0
>>> 0*AccumBounds(-1, 1)
0
>>> AccumBounds(-1, 1)*0
0
```

This is resulting in a regression in https://github.com/sympy/sympy/pull/13059 for `limit(atan(x)*sin(1/x), x, 0)` because the limit heuristics are going through the args in a different order. Probably an example could be found for master where an AccumBounds is returned from limit instead of 0.

### PR DESCRIPTION
**Pull request by asmeurer**
Fixes #13159.

### CODE REVIEW

**Comment by asmeurer** [comment]
@Upabjojr does this usage of postprocessors look correct to you?

File: sympy/calculus/util.py [hunk 1]
```diff
@@ -1223,5 +1224,32 @@ def union(self, other):
             return AccumBounds(other.min, Max(self.max, other.max))
 
 
+def _AccumulationBounds_constroctor_postprocessor_Mul(expr):
```
Comment (Upabjojr) [inline sympy/calculus/util.py]: typo?

**Comment by Upabjojr** [comment]
I don't see anything wrong.

**Comment by asmeurer** [comment]
Thanks. By the way, I originally wanted to make these static methods so they could be put right next to `__mul__` and `__add__`, but then I would still have to have the dictionary at the end, because you can't reference a class from its own body (AccumulationBounds would be undefined).

Maybe the metaclass could detect and add appropriately named static methods to the dictionary automatically (but then again, perhaps that would just slow down the import time for little benefit).

**Comment by Upabjojr** [comment]
One of my concerns about the post-processors is that the objects of the created expressions never get destroyed. There will always be a reference to them in the mapping dictionary.

**Comment by asmeurer** [comment]
Why is it necessary to have the objects in the mapping, along with the classes?

**Comment by Upabjojr** [comment]
> Why is it necessary to have the objects in the mapping, along with the classes?

You don't want to start the post-processors for every `Add` and `Mul`, only for those instances that actually contain the object of interest. The trick is to put the instances of those objects in the mapping.

Originally, there was an extra element in `__slots__`, but then I removed it after it had been labeled as _evil genius_.

**Comment by asmeurer** [comment]
Oh, I see, because you want it to work even for things that contain the object (like a Pow). 

Maybe we can use a [weak ref dictionary](https://docs.python.org/3/library/weakref.html#weakref.WeakKeyDictionary).

**Comment by Upabjojr** [comment]
> Oh, I see, because you want it to work even for things that contain the object (like a Pow).

Yes, otherwise `_op_priority` is enough.

Another alternative would be to store the tuple _(type, hash)_.

**Comment by asmeurer** [comment]
Honestly storing it on the object makes the most sense (and I'm not quite sure if that specific thing was what Ronan didn't like).

**Comment by Upabjojr** [comment]
The problem is that all objects would have a `dict` attached to them, which in most cases would just be empty.

**Comment by jksuom** [comment]
It seems that `is_finite` should be implemented for `AccumBounds`s. Otherwise all products with 0 are 0:
```
>>> Mul(0, AccumBounds(0, oo))
0
```
This should probably be `nan`, which would be returned by `Mul.flatten()` if `AccumBounds(0, oo).is_finite` were false.

**Comment by asmeurer** [comment]
There are some troubling test failures here.

**Comment by jksuom** [comment]
This leads to infinite recursion.
```
In [43]: a = AccumBounds(-1, 1)

In [44]: a*sqrt(x) + O(x)
```
It seems that the Add postprocessor classifies both terms as non_accum_args.

**Comment by asmeurer** [comment]
So the postprocessor itself needs to check which args actually contain AccumBounds.

**Comment by Upabjojr** [comment]
> So the postprocessor itself needs to check which args actually contain AccumBounds.

The unit consistency check works recursively backwards:
https://github.com/sympy/sympy/blob/master/sympy/physics/units/quantities.py#L206

Probably it's not very efficient.

**Comment by Upabjojr** [comment]
> So the postprocessor itself needs to check which args actually contain AccumBounds.

The unit consistency check works recursively backwards:
https://github.com/sympy/sympy/blob/master/sympy/physics/units/quantities.py#L206

Probably it's not very efficient.

**Comment by Upabjojr** [comment]
I believe this problem is more generic: one should make `Mul` behave like `*`.

**Comment by normalhuman** [comment]
But we probably don't want Mul(*args) to be literally a consecutive application of `*`, as this would be quadratic in the number of arguments (#13945).

**Comment by Upabjojr** [comment]
> But we probably don't want Mul(*args) to be literally a consecutive application of *, as this would be quadratic in the number of arguments.

We may still wish to find an optimized way to construct `Mul` and make the result look the same as a series of `*` operations.

**Comment by asmeurer** [comment]
For what it's worth, this PR has made me very uneasy about the postprocessors.  Basically, it seems very difficult to write the postprocessors in a way that's both efficient (doesn't do more passes over the args than necessary, not accidentally quadratic), and doesn't create memory leaks by keeping references to objects around forever. Maybe there's a way to salvage them, but my current thinking is that we should ditch them for something better.

There are some thoughts [here](https://github.com/sympy/sympy/pull/14490#issuecomment-384095425) on using multipledispatch instead.

**Comment by czgdp1807** [comment]
The associated issue has been fixed in https://github.com/sympy/sympy/pull/14957. Hence closing. Please feel free to re-open it if the issue isn't fixed completely. ping @asmeurer

**Pull request closed without merging**